File: guides/micronaut-microservices-distributed-tracing-jaeger-opentelemetry/bookinventory/java/src/main/java/example/micronaut/BooksController.java
package example.micronaut;

import io.micronaut.http.MediaType;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.Produces;
import io.micronaut.tracing.annotation.ContinueSpan;
import io.micronaut.tracing.annotation.SpanTag;

import javax.validation.constraints.NotBlank;
import java.util.Optional;

@Controller("/books")
public class BooksController {

    @Produces(MediaType.TEXT_PLAIN)
    @Get("/stock/{isbn}")
    @ContinueSpan // <1>
    public Boolean stock(@SpanTag("stock.isbn") @NotBlank String isbn) { // <2>
        return bookInventoryByIsbn(isbn).map(bi -> bi.getStock() > 0).orElse(null);
    }

    private Optional<BookInventory> bookInventoryByIsbn(String isbn) {
        if (isbn.equals("1491950358")) {
            return Optional.of(new BookInventory(isbn, 4));

        } else if (isbn.equals("1680502395")) {
            return Optional.of(new BookInventory(isbn, 0));
        }
        return Optional.empty();
    }
}

File: guides/micronaut-microservices-distributed-tracing-jaeger/bookinventory/python/src/example/micronaut/books_controller.py
from typing import Annotated

from jakarta.validation.constraints import NotBlank
from micronaut.http import MediaType
from micronaut.http.annotation import Get, Produces
from micronaut.tracing.annotation import ContinueSpan, SpanTag

from .book_inventory import BookInventory


@Produces(MediaType.TEXT_PLAIN)
@Get("/books/stock/{isbn}")
@ContinueSpan  # <1>
def stock(isbn: Annotated[str, SpanTag("stock.isbn"), NotBlank]) -> bool | None:  # <2>
    inventory = book_inventory_by_isbn(isbn)
    if inventory is None:
        return None
    return inventory.stock > 0


def book_inventory_by_isbn(isbn: str) -> BookInventory | None:
    if isbn == "1491950358":
        return BookInventory(isbn, 4)
    if isbn == "1680502395":
        return BookInventory(isbn, 0)
    return None
